core/server/master/src/test/java/alluxio/master/file/FileSystemMasterSyncMetadataTest.java

```diff
@@ -225,14 +244,84 @@ private AlluxioURI setupMockUfsS3Mount()
+    mFileSystemMaster.createDirectory(new AlluxioURI("/b/"), CreateDirectoryContext.defaults());
+    // If the sync operation happens, the flag will be marked
+    SyncAwareFileSystemMaster delegateMaster = (SyncAwareFileSystemMaster) mFileSystemMaster;
+    delegateMaster.setSynced(false);
+
+    delegateMaster.delete(new AlluxioURI("/a/"),
+        DeleteContext.mergeFrom(DeletePOptions.newBuilder()
+            .setRecursive(true).setAlluxioOnly(true)));
+    // The files have been deleted
+    assertEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/a/")));
+    assertEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/a/b/")));
+    // Irrelevant files are not affected
+    assertNotEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/b/")));
+    // Sync has not happened
+    assertFalse(delegateMaster.mSynced.get());
+  }
+
+  private static class SyncAwareFileSystemMaster extends DefaultFileSystemMaster {
+    AtomicBoolean mSynced = new AtomicBoolean(false);
+
+    public SyncAwareFileSystemMaster(BlockMaster blockMaster, CoreMasterContext masterContext,
+                                     ExecutorServiceFactory executorServiceFactory) {
+      super(blockMaster, masterContext, executorServiceFactory);
+    }
+
+    @Override
+    InodeSyncStream.SyncStatus syncMetadata(RpcContext rpcContext, AlluxioURI path,
+        FileSystemMasterCommonPOptions options, DescendantType syncDescendantType,
+        @Nullable FileSystemMasterAuditContext auditContext,
+        @Nullable Function<LockedInodePath, Inode> auditContextSrcInodeFunc,
+        @Nullable PermissionCheckFunction permissionCheckOperation,
+        boolean isGetFileInfo) throws AccessControlException, InvalidPathException {
+      mSynced.set(true);
+      return super.syncMetadata(rpcContext, path, options, syncDescendantType, auditContext,
+              auditContextSrcInodeFunc, permissionCheckOperation, isGetFileInfo);
+    }
+
+    void setSynced(boolean synced) {
+      mSynced.set(synced);
+    }
+  }
+
+  private class SyncAwareFileSystemMasterFactory implements MasterFactory<CoreMasterContext> {
+    @Override
+    public boolean isEnabled() {
+      return true;
+    }
+
+    @Override
+    public String getName() {
+      return "SyncAwareFileSystemMasterFactory";
+    }
+
+    @Override
+    public FileSystemMaster create(MasterRegistry registry, CoreMasterContext context) {
+      BlockMaster blockMaster = registry.get(BlockMaster.class);
+      FileSystemMaster fileSystemMaster = new SyncAwareFileSystemMaster(blockMaster, context,
+          ExecutorServiceFactories.constantExecutorServiceFactory(mExecutorService));
+      registry.add(FileSystemMaster.class, fileSystemMaster);
+      return fileSystemMaster;
+    }
+  }
+
   private void startServices() throws Exception {
+    mExecutorService = Executors
```

Comment:
Use a specified threadpool here instead of the default one from `DefaultFileSystemMaster` constructor